client.go

```diff
@@ -425,11 +430,22 @@ func (client *Client) cachedMetadata(topic string, partitionID int32) *Partition
 func (client *Client) cachedPartitions(topic string, partitionSet int) []int32 {
 	client.lock.RLock()
 	defer client.lock.RUnlock()
+	return client.cachedPartitionsNoLock(topic, partitionSet)
+}
 
+func (client *Client) cachedPartitionsNoLock(topic string, partitionSet int) []int32 {
 	partitions := client.metadata[topic]
+
 	if partitions == nil {
 		return nil
 	}
+	// Important:  If refactoring this, please keep the client.lock.RLock() above
+	previousResultMap, resMapExists := client.cachedPartitionsResults[topic]
+	if resMapExists {
+		if previousResult, exists := previousResultMap[partitionSet]; exists {
+			return previousResult
```

Comment:
This subtlety changes the behavior of this method by not returning a copy - if anyone has been relying on being able to edit the result their code will break.

I'm not sure if we need to care or not, but it's something to think about.